### ISSUE TITLE
Android WebView closing which results in an Transaction Error

### ISSUE THREAD

**Issue by codepopsicle** [citruspay/plug-play-sdk-android#2]
Hi,

I was integrating the Citrus Plug and Play SDK into my Android application. I have also hosted the BillGenerator and ReturnURL links on my server. The integration was successful and I proceeded to test using the sample BillGenerator and ReturnURLs given in the SDK. Everything worked fine with this. But when I replaced them with my own BillGenerator and ReturnURLs, the transaction was aborted when I clicked on any bank in net banking,; ie; basically when Web View was about to open.

We got the following error in the log as well "Attempt to remove non-JNI local reference, dumping thread". Can you help us with it please?

**Comment by salilgodbole** [comment]
Hi,

Could you please share the logs with us for further investigation?

Thanks,
*Salil Godbole*
Skype id: salilgodbole

**Comment by codepopsicle** [comment]
Hi Salil,

I'm attaching the file here.

[error_log.txt](https://github.com/citruspay/plug-play-sdk-android/files/146280/error_log.txt)

Also, we slightly suspect that its an issue with our BillGenerator code (in Python ) 
[billgen.txt](https://github.com/citruspay/plug-play-sdk-android/files/146284/billgen.txt)
as well. I am attaching it here.




Could you also confirm if using the Retrofit library, you're passing parameters when you call the BillGenerator URL using the GET method?

**Comment by salilgodbole** [comment]
Yes we are passing amount parameter with GET method to bill generator,

Thanks,
*Salil Godbole*
Skype id: salilgodbole

**Comment by salilgodbole** [comment]
Following is the format in which we call the bill generator url.
https://salty-plateau-1529.herokuapp.com/billGenerator.sandbox.php?amount=3.0

Thanks,
*Salil Godbole*
Skype id: salilgodbole

**Comment by codepopsicle** [comment]
Our URL will be called like this: http://whizzbuydev.herokuapp.com/billgen?amount=5. Does the amount need to be passed as a decimal?

Is there any way with the way we are returning the JSON response?

**Comment by salilgodbole** [comment]
I am getting application error while calling the url.
Could you check whether the url is working or not?


Thanks,
*Salil Godbole*
Skype id: salilgodbole

**Comment by codepopsicle** [comment]
Hi Salil,

Some code needs to be commented. Will sort it out ASAP and get back to you.

Regards,

Pritish

**Comment by codepopsicle** [comment]
Hi Salil,

Sorry for the delay in fixing the BillGenerator from our end. Now it works, you can check it out at:

http://whizzbuydev.herokuapp.com/billgenerator?amount=5

Now when we test, it seems that our BillGenerator is working and returning the correct response. But it is failing at the 3D secure page. The following are our observations while testing:

1) While using my own ICICI debit card, the following error was received.
![screenshot_1](https://cloud.githubusercontent.com/assets/2179528/13343775/ef9f4790-dc75-11e5-803d-6b1509435963.png)
![screenshot_2](https://cloud.githubusercontent.com/assets/2179528/13343776/efa55d56-dc75-11e5-87c8-e1fe1ac1abc0.png)


2) While using your HDFC test card as well my own HDFC credit card, I didn't get the above error, but got the attached response in the 3D secure page. 

![screenshot_3](https://cloud.githubusercontent.com/assets/2179528/13343811/4d3d1c10-dc76-11e5-90df-a03c6a3f034c.png)

The pgResponse code received from the response on the screen is 0, which indicates that it is a success. Why is this so since I haven't entered the 3D secure code or OTP?

Also, I have couple of queries to the Mobile Team:

1) In the Android Plug and Play SDK, how do I actually pass the amount and custom parameters while I am calling the BillGenerator? It seems that you are automatically concatenating the amount to the URL by default. Where can this be changed?

2) In the Android Plug and Play SDK, there seems to be a limit of Rs 5000 on the amount passed. Will this be the same in Production as well?

3) In the iOS SDK, do you have Swift version as well, or just an Objective-C one? If so, what support do you offer for iOS apps developed on Swift?


Kindly arrange a call to discuss the same today to ensure speedy resolution of these queries. And thanks for all your help so far.


Regards,

Pritish

**Comment by salilgodbole** [comment]
Hi Pritish,

The page with xml you see is a dummy 3d secure page. Since you are testing
on sandbox, the 3d secure authentication is not done and hence the
transaction is successful.
The page with html content you can see is your return url page. The
activity should get auto finished once the transaction is completed and you
should not see your return url page. To achieve this please make sure you
are calling CitrusResponse.pgResponse(data); from your return url, this
will capture the response in the SDK and SDK's screen will be auto-closed
and the control will be given back to your app.

Regarding the 5000 limit, I will get back to you on this.
For iOS support and further support, please drop an email to
*developer.support@citruspay.com
<developer.support@citruspay.com>* and mark *mobiledev@citruspay.com
<mobiledev@citruspay.com>* in cc.

Let me know if you need more information.

Thanks,
*Salil Godbole*
Skype id: salilgodbole

**Comment by codepopsicle** [comment]
Hi Salil,

Two more clarifications before I shift everything to email:

1) How exactly do I encode custom parameters in the BillGenerator URL when I'm calling it from your Plug and Play SDK? It seems you are attaching the '?amount=5' by default. And we need to pass custom amount and one internal transaction ID.

2) To call CitrusResponse.pgResponse(data) from my returnurl. Anyway you can give me an exmaple of how its done in Python? Your sample Python code doesn't seem to have it. Only your sample Java code has it.

Sorry if I'm inconveniencing you by posting my queries here. Thing is, we have to complete this urgently.

Regards,

Pritish

**Comment by salilgodbole** [comment]
Hi,

​​
Following is the sample response of bill generator. Url is
https://salty-plateau-1529.herokuapp.com/billGenerator.sandbox.php?amount=3.0
You can pass the custom parameters in following format.

{"merchantTxnId":"145646624922223","amount":{"value":"3.0","currency":"INR"},"requestSignature":"348c6eb530b87e8188475ac30d12f8f0fe63717f","merchantAccessKey":"F2VZD1HBS2VVXJPMWO77","returnUrl":"https:\/\/
salty-plateau-1529.herokuapp.com
\/redirectURL.sandbox.php","notifyUrl":"https:\/\/
salty-plateau-1529.herokuapp.com\/notifyUrl.sandbox.php","*customParameters*
":{"param1":"1000","param2":"CitrusTestSDK"},
"alteredAmout":{"value":null,"currency":"INR"},"dpSignature":"554beb58c77f208e03704ef961a8b36b34dbefc8"}

You may append your custom parameters to the bill url and pass that bill
url to the sdk.
e.g.
https://salty-plateau-1529.herokuapp.com/billGenerator.sandbox.php?param1=1000&param2=CitrusTestSDK
Now SDK will append amount to the url to form the request url like below
https://salty-plateau-1529.herokuapp.com/billGenerator.sandbox.php
?param1=1000&param2=CitrusTestSDK&amount=3.0

Also make sure you add the custom parameters in the Citrus Merchant Panel

2. You may refer following url for sample implementation of return url in
python.
http://www.citruspay.com/DevelopersGuide/android/returnurlAndroid.html#section-linetriangle-7

Let me know if you need more information.


Thanks,
*Salil Godbole*
Skype id: salilgodbole

**Comment by codepopsicle** [comment]
Hi Salil,

Thanks a lot for clearing my doubts. I will try these out and get back to you.

But I'm not clear where on my Merchant Panel I'm supposed to be setting custom parameters?

Could you guide me? If my understanding is correct, whatever custom parameters I include in the BillGenerator URL, I will have to include in my Merchant Panel?

Regards,

Pritish

**Comment by codepopsicle** [comment]
Hi Salil,

I'm not able to call the CitrusResponse.pgResponse(data) from my ReturnURL as it is in Python and the CitrusResponse call is in Javascript I think. Would it be possible to put you in touch with someone from your Python team?

I have
[returnurl.txt](https://github.com/citruspay/plug-play-sdk-android/files/149215/returnurl.txt)

 attached my current ReturnURL code

Regards,

Pritish

**Comment by sathishravula** [comment]
How to pass Order Id to bill generator from android plug and play SDK

**Comment by VenkateswarluAtSipani** [comment]
Hi,

I am trying append custom parameters to the bill url and pass that bill
url to the sdk.
e.g.
https://salty-plateau-1529.herokuapp.com/billGenerator.sandbox.php?id=10

In Android CitrusPaySDK on doing this way the amount parameter is appended to the url as 
https://salty-plateau-1529.herokuapp.com/billGenerator.sandbox.php
?id=10&amount=3.00

But iOS SDK is not checking whether there is already a parameter in url and it appends amount to the url to form the request url like below
https://salty-plateau-1529.herokuapp.com/billGenerator.sandbox.php
?id=10?amount=3.00

**Comment by MangeshKadam** [comment]
Hi @VenkateswarluAtSipani  - I can not find any difference in billURL in andorid and iOS as you mentioned in your comment above. Both looks similar. Can you recheck?

**Comment by VenkateswarluAtSipani** [comment]
Observe both the urls you can find "?amount=3.00" in iOS  and "&amount=3.00" in Android

**Comment by MangeshKadam** [comment]
can you raise this issue @ https://github.com/citruspay/citruspay-ios-sdk?

**Comment by VenkateswarluAtSipani** [comment]
Issue is already there.
